cdemu-daemon/man/cdemu-daemon.8
.TH CDEMU-DAEMON 8 "Nov 19, 2025"
.SH NAME
cdemud \- The CDEmu daemon
.SH SYNOPSIS
.B cdemu-daemon
[options]
.SH DESCRIPTION
This is CDEmu daemon, the userspace part of the cdemu suite, a
free, GPL CD/DVD-ROM device emulator for linux.

It receives SCSI commands from kernel module and processes them, passing the
requested data back to the kernel. Daemon implements the actual virtual device;
one instance per each device registered by kernel module. It uses libMirage, an
image access library that is part of the cdemu suite, for the image access
(e.g. sector reading).

The daemon registers itself on D-BUS' system or session bus (depending on the
options passed to it) where it exposes an interface that can be used by clients
to control it.
.SH OPTIONS
.TP
.B --config-file=filename
Configuration file to load. See \fBConfiguration file\fR section for details.
.TP
.B -n --num-devices=N
Number of devices. Specifies the number of virtual devices the daemon instance should
register with kernel module and control. By default, 1 virtual device is created.
.TP
.B -c --ctl-device=path
Control device. Specifies the control device path. Control device is a character device
provided by kernel module; it is used for communication between kernel and userspace
part of CDEmu suite. By default, /dev/vhba_ctl is used.
.TP
.B -a --audio-driver=driver
Audio driver. Specifies audio driver to be used by virtual devices' audio play.
Since libao is used for audio playback, any valid libao driver can be specified.
If \fBdefault\fR is used as a driver, then libao's default driver is used. Few
examples of valid libao drivers are \fBalsa\fR, \fBoss\fR, \fBpulse\fR or \fBnull\fR.
For all drivers, default device is used. By default, if audio driver is not explicitly
specified via this option, \fBnull\fR is used.
.TP
.B -b --bus=bus_type
Bus type. Valid options are \fBsystem\fR or \fBsession\fR. By default, session bus is used.
(Please note that the running of CDEmu daemon on system bus is considered
deprecated and is discouraged for security reasons).
.TP
.B -l --logfile=logfile
Log file to write logging output into. By default, use of log file is disabled and messages
are written to stdout.
.TP
.B --default-cdemu-debug-mask=mask
Default debug mask for CDEmu devices (at CDEmu daemon implementation level).
Default value is 0.
.TP
.B --default-mirage-debug-mask=mask
Default debug mask for CDEmu devices (at libMirage implementation level).
Default value is 0.
.TP
.B --system-sleep-handler=0|1
Enable or disable systemd-logind system sleep handler. When enabled,
the daemon attempts to obtain delaying sleep inhibitor lock from the
\fIorg.freedesktop.login1.Manager\fR systemd-logind interface at
\fI/org/freedesktop/login1\fR on the system bus, and registers handler
for the \fIprepare-for-sleep\fR signal. When system is about to enter
suspend/hibernation (as indicated by received signal), the virtual devices
are stopped, and inhibitor lock is released. After system resumes (as
indicated by received signal), the inhibitor lock is re-acquired, and
virtual devices are restarted. This mechanism attempts to mitigate issues
with active virtual devices preventing the system from properly
suspending/hibernating. Default value is 1 (the handler is enabled). If
sleep handler functionality is disabled at build time, this option has
no effect.
.TP
.B -? --help
Display the help message.
.SH EXAMPLES
.TP
.B Running daemon on default bus with PulseAudio audio driver and 4 virtual devices:
cdemu-daemon --num-devices=4 --audio-driver=pulse
.TP
.B Running daemon on system bus with 2 virtual devices and logging redirected to a file:
cdemu-daemon --num-devices=2 --bus=system --logfile=/tmp/cdemu-daemon.log
.SH CONFIGURATION FILE
The daemon's settings can be collected in a configuration file, which
is passed to the daemon via \fB--config-file\fR option. Configuration
file contains \fBkey=value\fR pairs that are contained in a single
\fB[settings]\fR section. The keys in configuration file correspond to
the long names of command-line switches.

If an option is specified in both configuration file and on command-line,
the latter takes precedence.

If the specified configuration file does not exist, it is silently ignored.
If the specified configuration file is invalid (e.g., has invalid syntax,
or does not contain a section), an error is raised.

.B Example

Saving the following configuration to a file and passing it to the daemon
via \fB--config-file\fR is equivalent to passing \fB--num-devices=2
--audio-driver=pulse\fR directly to the daemon:

.RS\fB
[settings]
.br
num-devices=2
.br
audio-driver=pulse
\fR
.RE

.SH AUTHORS
.PP
Rok Mandeljc <rok.mandeljc@gmail.com>
.PP
CDEmu project's web page: http://cdemu.sourceforge.net
.PP
CDEmu project's mailing list: cdemu-devel@lists.sourceforge.net


ChangeLog
==========================================================================
 Version 1.2.0   2009-07-26
==========================================================================

Rok Mandeljc:
	* Adaptations to new libmirage parser framework.
	* Password support.
	* Versioning support.
	* Updated Slovenian translation.
	* Updated documentation.
	* Misc. fixes.

Loïc Martin:
	* French translation.

==========================================================================
 Version 1.1.0   2008-06-29
==========================================================================

Marcel Hoppe:
	* German translation.

Rok Mandeljc:
	* Added documentation.
	* Slovenian translation.
	* Added 'any' and 'all' keywords.
	* Many improvements to the code.
	* Added reporting of device mappings.
	* Many bugfixes.

Henrik Stokseth:
	* Norwegian translation.
	* Code improvements and bugfixes.

==========================================================================
 Version 1.0.0   2007-12-24
==========================================================================

Rok Mandeljc:
	* Initially wrote the whole CDEmu client.



cdemu-daemon/src/cdemu-daemon.c
/*
 *  CDEmu daemon: Daemon object
 *  Copyright (C) 2006-2012 Rok Mandeljc
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "cdemu.h"
#include "cdemu-daemon-private.h"

#define __debug__ "Daemon"


/**********************************************************************\
 *                           Signal handlers                          *
\**********************************************************************/
static void device_status_changed_handler (CdemuDevice *device, CdemuDaemon *self)
{
    gint number = cdemu_device_get_device_number(device);
    cdemu_daemon_dbus_emit_device_status_changed(self, number);
}

static void device_option_changed_handler (CdemuDevice *device, gchar *option, CdemuDaemon *self)
{
    gint number = cdemu_device_get_device_number(device);
    cdemu_daemon_dbus_emit_device_option_changed(self, number, option);
}

static void device_mapping_ready_handler (CdemuDevice *device, CdemuDaemon *self)
{
    gint number = cdemu_device_get_device_number(device);
    cdemu_daemon_dbus_emit_device_mapping_ready(self, number);
}


/**********************************************************************\
 *                     Device restart on inactivity                   *
\**********************************************************************/
struct DaemonDevicePtr {
    CdemuDaemon *daemon;
    CdemuDevice *device;
};

static gboolean device_restart_callback (struct DaemonDevicePtr *data)
{
    CdemuDaemon *self = data->daemon;
    CdemuDevice *device = data->device;

    /* Properly stop the device (I/O thread clenup) */
    cdemu_device_stop(device);

    /* Start the device */
    if (!cdemu_device_start(device, self->priv->ctl_device)) {
        CDEMU_DEBUG(device, DAEMON_DEBUG_WARNING, "%s: failed to restart device!\n", __debug__);
    } else {
        CDEMU_DEBUG(device, DAEMON_DEBUG_DEVICE, "%s: device started successfully\n", __debug__);
    }

    /* Free the pointer structure */
    g_free(data);

    return FALSE;
}

/* The signal handler; since the signal is emitted from the device's
   I/O thread, this handler is also executed there... so we need to get
   into our main thread first, which is done by scheduling an idle function */
static void device_kernel_io_error_handler (CdemuDevice *device, CdemuDaemon *self)
{
    struct DaemonDevicePtr *data = g_new(struct DaemonDevicePtr, 1);
    data->daemon = self;
    data->device = device;

    CDEMU_DEBUG(device, DAEMON_DEBUG_DEVICE, "%s: restarting device...\n", __debug__);
    g_idle_add((GSourceFunc)device_restart_callback, data);
}


/******************************************************************************\
 *                                 Public API                                 *
\******************************************************************************/
gboolean cdemu_daemon_initialize_and_start (CdemuDaemon *self, gint num_devices, gchar *ctl_device, gchar *audio_driver, gboolean system_bus)
{
    MirageContext *context;
    GBusType bus_type = system_bus ? G_BUS_TYPE_SYSTEM : G_BUS_TYPE_SESSION;

    self->priv->ctl_device = g_strdup(ctl_device);
    self->priv->audio_driver = g_strdup(audio_driver);

    /* Create a MirageContext and use it as debug context */
    context = g_object_new(MIRAGE_TYPE_CONTEXT, NULL);
    mirage_context_set_debug_name(context, "cdemu");
    mirage_context_set_debug_domain(context, "CDEMU");
    mirage_contextual_set_context(MIRAGE_CONTEXTUAL(self), context);
    g_object_unref(context);

    /* Glib's main loop */
    self->priv->main_loop = g_main_loop_new(NULL, FALSE);

    /* Initialize libao */
    ao_initialize();

    /* Make sure the D-BUS name we're going to use isn't taken already (by another
       instance of the server). We're actually going to claim it once we create
       the devices, but we want to avoid the device creation if name claim is
       already doomed to fail... */
    if (!cdemu_daemon_dbus_check_if_name_is_available(self, bus_type)) {
        return FALSE;
    }

    /* Create desired number of devices */
    for (gint i = 0; i < num_devices; i++) {
        if (!cdemu_daemon_add_device(self)) {
            CDEMU_DEBUG(self, DAEMON_DEBUG_WARNING, "%s: failed to create device!\n", __debug__);
            return FALSE;
        }
    }

    /* Register on D-Bus bus */
    cdemu_daemon_dbus_register_on_bus(self, bus_type);

    /* Run the main loop */
    g_main_loop_run(self->priv->main_loop);


    /* Cleanup D-Bus */
    cdemu_daemon_dbus_cleanup(self);

    return TRUE;
}

void cdemu_daemon_stop_daemon (CdemuDaemon *self)
{
    /* Stop the main loop */
    g_main_loop_quit(self->priv->main_loop);
}


/**********************************************************************\
 *                          Device management                         *
\**********************************************************************/
gboolean cdemu_daemon_add_device (CdemuDaemon *self)
{
    CdemuDevice *device;
    gint device_number;

    device_number = g_list_length(self->priv->devices);

    /* Create and initialize device object */
    device = g_object_new(CDEMU_TYPE_DEVICE, NULL);
    if (!cdemu_device_initialize(device, device_number, self->priv->audio_driver)) {
        CDEMU_DEBUG(self, DAEMON_DEBUG_WARNING, "%s: failed to initialize device #%i\n", __debug__, device_number);
        g_object_unref(device);
        return FALSE;
    }

    /* Don't set parent, as devices have their own debug contexts */

    /* Add handling for signals from the device... this allows us to
       pass them on via DBUS */
    g_signal_connect(device, "status-changed", (GCallback)device_status_changed_handler, self);
    g_signal_connect(device, "option-changed", (GCallback)device_option_changed_handler, self);
    g_signal_connect(device, "kernel-io-error", (GCallback)device_kernel_io_error_handler, self);
    g_signal_connect(device, "mapping-ready", (GCallback)device_mapping_ready_handler, self);

    /* Start device */
    if (!cdemu_device_start(device, self->priv->ctl_device)) {
        CDEMU_DEBUG(self, DAEMON_DEBUG_WARNING, "%s: failed to start device #%i!\n", __debug__, device_number);
        g_object_unref(device);
        return FALSE;
    }

    /* Add it to devices list */
    self->priv->devices = g_list_append(self->priv->devices, device);

    /* Emit signal */
    cdemu_daemon_dbus_emit_device_added(self);

    return TRUE;
}

gboolean cdemu_daemon_remove_device (CdemuDaemon *self)
{
    /* Get last device entry */
    GList *entry = g_list_last(self->priv->devices);
    if (!entry) {
        return FALSE;
    }

    /* Release the device, which is enough to stop and free it */
    g_object_unref(entry->data);

    /* Remove and free node from the list */
    self->priv->devices = g_list_delete_link(self->priv->devices, entry);

    /* Emit signal */
    cdemu_daemon_dbus_emit_device_removed(self);

    return TRUE;
}

CdemuDevice *cdemu_daemon_get_device (CdemuDaemon *self, gint device_number, GError **error)
{
    CdemuDevice *device = g_list_nth_data(self->priv->devices, device_number);
    if (!device) {
        g_set_error(error, CDEMU_ERROR, CDEMU_ERROR_INVALID_ARGUMENT, "Invalid device number!");
        return NULL;
    }
    return g_object_ref(device);
}


/**********************************************************************\
 *                             Object init                            *
\**********************************************************************/
G_DEFINE_TYPE(CdemuDaemon, cdemu_daemon, MIRAGE_TYPE_OBJECT);

static void cdemu_daemon_init (CdemuDaemon *self)
{
    self->priv = CDEMU_DAEMON_GET_PRIVATE(self);

    self->priv->main_loop = NULL;
    self->priv->devices = NULL;
    self->priv->ctl_device = NULL;
    self->priv->audio_driver = NULL;

    /* Set version string */
    self->priv->version = g_strdup(CDEMU_DAEMON_VERSION);

    /* D-Bus data */
    self->priv->connection = NULL;
    self->priv->owner_id = 0;
}

static void cdemu_daemon_dispose (GObject *gobject)
{
    CdemuDaemon *self = CDEMU_DAEMON(gobject);

    /* Unref main loop */
    g_main_loop_unref(self->priv->main_loop);

    /* Unref all devices */
    for (GList *entry = self->priv->devices; entry; entry = entry->next) {
        CdemuDevice *dev = entry->data;
        if (dev) {
            g_object_unref(dev);
            entry->data = NULL;
        }
    }

    /* Chain up to the parent class */
    return G_OBJECT_CLASS(cdemu_daemon_parent_class)->dispose(gobject);
}

static void cdemu_daemon_finalize (GObject *gobject)
{
    CdemuDaemon *self = CDEMU_DAEMON(gobject);

    g_free(self->priv->ctl_device);
    g_free(self->priv->audio_driver);

    /* Free devices list */
    g_list_free(self->priv->devices);

    /* Free version string */
    g_free(self->priv->version);

    /* Shutdown libao */
    ao_shutdown();

    /* Chain up to the parent class */
    return G_OBJECT_CLASS(cdemu_daemon_parent_class)->finalize(gobject);
}

static void cdemu_daemon_class_init (CdemuDaemonClass *klass)
{
    GObjectClass *gobject_class = G_OBJECT_CLASS(klass);

    gobject_class->dispose = cdemu_daemon_dispose;
    gobject_class->finalize = cdemu_daemon_finalize;

    /* Register private structure */
    g_type_class_add_private(klass, sizeof(CdemuDaemonPrivate));
}
